src/main/java/com/maxmind/db/SingleBuffer.java

```diff
@@ -0,0 +1,174 @@
+import java.nio.channels.FileChannel.MapMode;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.CharsetDecoder;
+
+/**
+ * A {@link Buffer} implementation backed by a single {@link ByteBuffer}.
+ *
+ * <p>This implementation is limited to capacities up to
+ * {@link Integer#MAX_VALUE}, as {@link ByteBuffer} cannot exceed that size.
+ */
+public class SingleBuffer implements Buffer {
+
+    private final ByteBuffer buffer;
+
+    /**
+     * Creates a new {@code SingleBuffer} with the given capacity.
+     *
+     * @param capacity the capacity in bytes (must be <= Integer.MAX_VALUE)
+     * @throws IllegalArgumentException if the capacity exceeds
+     *                                  {@link Integer#MAX_VALUE}
+     */
+    public SingleBuffer(long capacity) {
+        if (capacity > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException(
+                "SingleBuffer cannot exceed Integer.MAX_VALUE capacity"
+            );
+        }
+        this.buffer = ByteBuffer.allocate((int) capacity);
+    }
+
+    /**
+     * Creates a new {@code SingleBuffer} wrapping the given {@link ByteBuffer}.
+     *
+     * @param buffer the underlying buffer
+     */
+    private SingleBuffer(ByteBuffer buffer) {
+        this.buffer = buffer;
+    }
+
+    /**
+     * Validates and converts a long index to an int for use with
+     * {@link ByteBuffer}.
+     *
+     * @param index the index to check
+     * @return the index as an int
+     * @throws IndexOutOfBoundsException if the index is negative or exceeds
+     *                                   {@link Integer#MAX_VALUE}
+     */
+    private int checkIndex(long index) {
+        if (index < 0 || index > Integer.MAX_VALUE) {
+            throw new IndexOutOfBoundsException(
+                "Index out of bounds for SingleBuffer: " + index
+            );
+        }
+        return (int) index;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long capacity() {
+        return buffer.capacity();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long position() {
+        return buffer.position();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer position(long newPosition) {
+        buffer.position(checkIndex(newPosition));
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long limit() {
+        return buffer.limit();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer limit(long newLimit) {
+        buffer.limit(checkIndex(newLimit));
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public byte get() {
+        return buffer.get();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer get(byte[] dst) {
+        buffer.get(dst);
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public byte get(long index) {
+        return buffer.get(checkIndex(index));
```

Comment:
Having a bounds check on ever `get(long)` is going to add up in terms of performance. We should already have sufficient checks to ensure that the index does not exceed the size of the database in the reader and decoder code. I think we could drop most uses of `checkIndex`. A safety check in the constructor does make sense though.